### ISSUE TITLE
How to avoid triggering unrelated jobs?

### ISSUE THREAD

**Issue by vince166** [jenkinsci/generic-webhook-trigger-plugin#14]
I have a single Jenkins instance hosting multiple jobs, which build different repos. I noticed a post message from one repo would trigger unrelated jobs. I understand that there is only one webhook url, but shouldn't Jenkins do some internal matching to trigger only the job that is relevant? For example, with the GitHub Plugin, I didn't see unrelated job being triggered. But that is not the case with the Generic Webhook Trigger Plugin. 

Is it by design, or am I missing something? 

If this is how it was designed, can we have multiple "Optional Filters", or can we have "positive" "Parameter Filter"? (I understand we can extract the source repo as a parameter, but the regex can only remove the match, not retain the match; I could use the "Optional Fitler" for matching the repo, but I am already using it to match a tag). 

This is an awesome plugin. Thank you for the great work!

**Comment by vince166** [comment]
btw, we are using the plugin with pipeline jobs (the plugin was configured under the "Build Triggers" section).

**Comment by tomasbjerre** [comment]
I would configure a variable to resolve to the repo name. And then ad an optional filter with that variable and with regexp like:
https://jex.im/regulex/#!embed=false&flags=&re=%5E(repo1%7Crepo2%7Crepo3)%24

You can have multiple variables in the optional regexp filter. Like `$TAGNAME $REPONAME`. And then you may want a regexp like:
https://jex.im/regulex/#!embed=false&flags=&re=%5ETHETAGNAME-%5B0-9%5D%2B%5Cs(repo1%7Crepo2%7Crepo3)%24

To match tags like THETAGNAME-2 in one of repo1, repo2 or repo3.

**Comment by vince166** [comment]
Thank you for the quick reply. Good tip for including two variables in the option filter. I will use that.

**Comment by xuliugen** [comment]
1、Git Post Data：

```
{
  "before": "1848fbfb6ae15769e6b31e9c4477d8150b018453",
  "after": "5cab18a38eaa83240ab86c7b775a9b27b51ef11d",
  "ref": "refs/heads/develop",
  "user_name": "\u5f90\u5218\u6839",
  "user": {
    "email": "1573876303@qq.com",
    "name": "\u5f90\u5218\u6839",
    "time": "2017-07-27T18:43:16+08:00"
  },
  "repository": {
    "name": "jenkins-test",
    "url": "https://git.oschina.net/xuliugen/jenkins-test",
    "description": "jenkins-test",
    "homepage": "https://git.oschina.net/xuliugen/jenkins-test"
  },
  "commits": [
    {
      "id": "5cab18a38eaa83240ab86c7b775a9b27b51ef11d",
      "message": "\u66f4\u6539\u4ee3\u7801\u66f4\u6539\u4ee3\u7801",
      "timestamp": "2017-07-27T18:43:11+08:00",
      "url": "https://git.oschina.net/xuliugen/jenkins-test/commit/5cab18a38eaa83240ab86c7b775a9b27b51ef11d",
      "author": {
        "name": "xuliugen",
        "email": "1573876303@qq.com",
        "time": "2017-07-27T18:43:11+08:00"
      }
    }
  ],
  "total_commits_count": 1,
  "commits_more_than_ten": false,
  "project": {
    "name": "jenkins-test",
    "path": "jenkins-test",
    "url": "https://git.oschina.net/xuliugen/jenkins-test",
    "git_ssh_url": "git@git.oschina.net:xuliugen/jenkins-test.git",
    "git_http_url": "https://git.oschina.net/xuliugen/jenkins-test.git",
    "git_svn_url": "svn://git.oschina.net/xuliugen/jenkins-test",
    "namespace": "xuliugen",
    "path_with_namespace": "xuliugen/jenkins-test",
    "default_branch": "master"
  },
  "hook_name": "push_hooks",
  "password": "admin"
}
```

2、I want to specify the branch of ‘refs/heads/develop’， And the Identifier is ‘ref’ ：

>  "ref": "refs/heads/develop",

3、The setting as follow：

![image](https://user-images.githubusercontent.com/12216664/28667174-f44d92f0-72fc-11e7-92a6-934937a8d8d7.png)

4、Jenkins log：

```
27-Jul-2017 18:59:02.385 INFO [Handling POST /generic-webhook-trigger/invoke from 61.160.47.7 : http-apr-8020-exec-1] org.jenkinsci.plugins.gwt.GenericWebHookRequestReceiver.doInvoke Triggering jenkins-test
27-Jul-2017 19:00:08.530 INFO [Executor #1 for master : executing jenkins-test #21] hudson.model.Run.execute jenkins-test #21 main build action completed: SUCCESS
27-Jul-2017 19:01:04.398 INFO [Handling POST /generic-webhook-trigger/invoke from 61.160.47.7 : http-apr-8020-exec-8] org.jenkinsci.plugins.gwt.GenericWebHookRequestReceiver.doInvoke Triggering jenkins-test
27-Jul-2017 19:01:04.400 INFO [Handling POST /generic-webhook-trigger/invoke from 61.160.47.7 : http-apr-8020-exec-8] org.jenkinsci.plugins.gwt.GenericTrigger.isMatching Not triggering "^(refs/heads/develop)$" not matching "refs/heads/feature/liugen.xu/20170727-webhooks".

```

5、It work！

**Comment by ajaypandit** [comment]
I am trying to set a dynamic filter based on the repo and the job name (we use same Jenkinsfile in all the repos) with following settings:

regexpFilterText: '$repo',   
regexpFilterExpression: '/${env.JOB_NAME}.split("/")[1]/g' 

The $repo is correctly set to the repo name e.g. template-service and I want to match it with the repo name in the job e.g. ORG/template-service/master. Even with this filter, the webhook is triggering all the jobs. I tried changing '${env.JOB_NAME}' to 'env.JOB_NAME' but that does not work either. Is it possible to use environment variables in regexpFilterExpression? Why is this filter triggering all the jobs?

**Comment by forpix** [comment]
In Jenkins I have a job which is configured to trigger on web-hook with GitHub, In GitHub I have 2 branches one is master and other is working branch. when I committed on master Jenkins trigger the job which is perfectly all right. But when I tried to pull request to working branch from master for the committed changes again a new build getting triggering which is completely annoying.
so is there any logic to stop the triggering the job for only pull-request....
